Source/CombatExtended/Harmony/Harmony_StunHandler.cs

```diff
@@ -12,86 +13,86 @@ namespace CombatExtended.HarmonyCE
-			int adaptationReduction = Mathf.RoundToInt(Mathf.Sqrt(dinfo.Amount * 45));
-			
-			if (adaptationReduction < ___EMPAdaptedTicksLeft) {
-			    ___EMPAdaptedTicksLeft -= adaptationReduction;
-			}
-			else
-			{
-			    float adaptationReductionRatio = (adaptationReduction - ___EMPAdaptedTicksLeft) / adaptationReduction;
-			    newStunAdaptedTicks = Mathf.RoundToInt(newStunAdaptedTicks * adaptationReductionRatio);
-			    newStunTicks = Mathf.RoundToInt(newStunTicks * adaptationReductionRatio);
-			    reStun();
-			}
-		    }
-		    		    
-		}
-		else
-		{
-		    __instance.StunFor_NewTmp(Mathf.RoundToInt(dinfo.Amount * 30f), dinfo.Instigator, true, true);
-		    ___EMPAdaptedTicksLeft = Mathf.RoundToInt(dinfo.Amount * 45 * bodySize);
-		    ___stunFromEMP = true;
-		    
-		}
-	    }
-	    return true;
-	}
+        public static bool Prefix(StunHandler __instance, DamageInfo dinfo, bool affectedByEMP, int ___EMPAdaptedTicksLeft, int ___stunTicksLeft, bool ___stunFromEMP)
+        {
+            float bodysize = BUILDING_DEFAULT_BODYSIZE;
+            if (__instance.parent is Pawn)
+            {
+                Pawn p = (Pawn)__instance.parent;
+                if (p.Downed || p.Dead) { return false; }
+                bodysize = p.BodySize;
+            }
+
+            Type t = typeof(StunHandler);
+            FieldInfo EMPAdaptedTicksLeft = t.GetField("EMPAdaptedTicksLeft", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance);
+            FieldInfo StunTicksLeft = t.GetField("stunTicksLeft", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance);
+            FieldInfo StunFromEMP = t.GetField("stunFromEMP", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance);
+            
+            float multiplier = bodysize / Mathf.Pow(bodysize, 2); // Multiplier based on bodysize, smaller for larger creatures
+            if (dinfo.Def == DamageDefOf.EMP && affectedByEMP)
+            {
+                if(___EMPAdaptedTicksLeft > 0)
+                {
+                    
+                    if (___stunTicksLeft > 0) 
+                    { EMPAdaptedTicksLeft.SetValue(__instance, ___EMPAdaptedTicksLeft - (int)(dinfo.Amount * FLAT_ADAPTATION_TIME_REDUCTION_RATIO * REDUCTION_MULTIPLIER_WHILE_STUNNED * multiplier)); }
+                    else 
+                    { EMPAdaptedTicksLeft.SetValue(__instance, ___EMPAdaptedTicksLeft - (int)(dinfo.Amount * FLAT_ADAPTATION_TIME_REDUCTION_RATIO * multiplier)); }
+                    
+                    
+                    if(___EMPAdaptedTicksLeft < 0)
+                    {
+                        float stun_time = Mathf.Abs(___EMPAdaptedTicksLeft / (ADAPTATION_TIME_RATIO / STUN_TIME_DAMAGE_RATIO));
```

Comment:
already know that `___EMPAdaptedTicksLeft < 0`, and that the denominator is positive (with current constants the value is `3` ). Can change to
`float stun_time = -( ___EMPAdaptedTicksLeft / ( ADAPTATION_TIME_RATIO / STUN_TIME_DAMAGE_RATIO );` for the same effect without calling `Mathf.Abs`
```suggestion
                        float stun_time = -(___EMPAdaptedTicksLeft / (ADAPTATION_TIME_RATIO / STUN_TIME_DAMAGE_RATIO));
```